src/explore-education-statistics-common/src/modules/table-tool/utils/applyTableHeadersOrder.ts

```diff
@@ -0,0 +1,342 @@
+import { TableHeadersConfig } from '@common/modules/table-tool/types/tableHeaders';
+import {
+  CategoryFilter,
+  LocationFilter,
+  TimePeriodFilter,
+  Filter,
+} from '@common/modules/table-tool/types/filters';
+import isEqual from 'lodash/isEqual';
+import intersectionWith from 'lodash/intersectionWith';
+import intersection from 'lodash/intersection';
+
+/**
+ * Apply previous ordering to default table headers.
+ */
+const applyTableHeadersOrder = ({
+  reorderedTableHeaders,
+  defaultTableHeaders,
+}: {
+  reorderedTableHeaders: TableHeadersConfig;
+  defaultTableHeaders: TableHeadersConfig;
+}): TableHeadersConfig => {
+  const defaultRowGroups = [
+    ...defaultTableHeaders.rowGroups,
+    defaultTableHeaders.rows,
+  ];
+  const defaultColGroups = [
+    ...defaultTableHeaders.columnGroups,
+    defaultTableHeaders.columns,
+  ];
+  const reorderedRowGroups = [
+    ...reorderedTableHeaders.rowGroups,
+    reorderedTableHeaders.rows,
+  ];
+  const reorderedColGroups = [
+    ...reorderedTableHeaders.columnGroups,
+    reorderedTableHeaders.columns,
+  ];
+
+  const { orderedGroups: orderedRowGroups } = orderGroups({
+    defaultGroups: defaultRowGroups,
+    reorderedGroups: reorderedRowGroups,
+    otherAxisDefaultGroups: defaultColGroups,
+    otherAxisReorderedGroups: reorderedColGroups,
+  });
+
+  const moveNewColumnGroupsToRows = orderedRowGroups.length <= 1;
+
+  const {
+    orderedGroups: allOrderedColumnGroups,
+    newGroups: newColumnGroups,
+  } = orderGroups({
+    defaultGroups: defaultColGroups,
+    reorderedGroups: reorderedColGroups,
+    otherAxisDefaultGroups: defaultRowGroups,
+    otherAxisReorderedGroups: reorderedRowGroups,
+    returnNewGroups: moveNewColumnGroupsToRows,
+  });
+
+  const allOrderedColumnGroupsLength = allOrderedColumnGroups.length;
+
+  // If there's zero or one row groups then add any new column groups
+  // to row groups instead to avoid causing horizontal scrolling
+  // by having lots of columns.
+  const allOrderedRowGroups = moveNewColumnGroupsToRows
+    ? [...orderedRowGroups, ...newColumnGroups]
+    : orderedRowGroups;
+  const allOrderedRowGroupsLength = allOrderedRowGroups.length;
```

Comment:
Feels like this heuristic only covers a really small number of use cases. If the user reorders to have say, 4 column groups and 2 row groups, we could still end up adding more column groups.

Maybe we could do something like the following instead:

- Extract the new groups for both rows and cols.
- Use `getSortedRowColGroups` function from `getDefaultTableHeadersConfig.ts` on these new groups. We should get the 'optimised' rows and col groups.
- Append these groups to the existing reordered row/col groups as appropriate

Think this would be a little bit more robust and would hopefully cover more cases!

Reply:
I'm not sure that'd be right either - just quickly trying and running the tests resulted in new row groups being distributed into rows and columns, so adding more columns than there would have been previously.  

I'll have a think about how else to do it, but not sure how much more time to put into this, as in the scenario where a user has put lots of groups in columns then adds more they can move them to rows if it's causing the table to be too wide.

Reply:
Ah yes of course!

I'd suggest writing up a simple function that fills up the row/cols, with a preference to populating the row groups first. Maybe something like:

```js
function appendNewGroups({
  rowGroups,
  colGroups,
  newGroups
}) {
  const nextRowGroups = [...rowGroups];
  const nextColGroups = [...colGroups];

  // Do a pre-sort so the bigger groups with with more options 
  // and wordier labels get placed first
  const sortedNewGroups = orderBy(newGroups, [
    options => options.length,
    options => options.reduce((acc, option) => acc + option.label.length, 0),
  ]);

  sortedNewGroups.forEach((group) => {
    if (nextRowGroup.length > nextColGroup.length) {
      nextColGroups.push(group);
    } else {
      nextRowGroups.push(group);
    }
  });

  return {
    rowGroups: nextRowGroups,
    colGroups: nextColGroups,
  };
}
```

Reply:
Cool, I had a quick go and that looks like it'll work, thanks. Will finish off when back on dfe.